llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/SplitReduction.cpp

```diff
@@ -0,0 +1,319 @@
+  ShapedType valuesOrigType = valuesOrig.getType().cast<ShapedType>();
+  Type valueElementType = valuesOrigType.getElementType();
+  Type indicesElementType =
+      topkOp.getResultTypes()[1].cast<ShapedType>().getElementType();
+
+  SmallVector<int64_t> expandedShape = getExpandedShape(
+      valuesOrigType.getShape(), splitReductionRatio, splitDimParallel);
+  RankedTensorType valuesExpandedType =
+      RankedTensorType::get(expandedShape, valueElementType);
+
+  // Expand input values shape for parallel processing
+  Value valuesExpanded = rewriter.create<tensor::ExpandShapeOp>(
+      loc, valuesExpandedType, valuesOrig, reassociationIndices);
+
+  // Define the expanded output types
+  SmallVector<int64_t> expandedResultShape = expandedShape;
+  expandedResultShape[kDimParallel] = kSize;
+  RankedTensorType outputValuesExpandedType =
+      RankedTensorType::get(expandedResultShape, valueElementType);
+  RankedTensorType outputIndicesExpandedType =
+      RankedTensorType::get(expandedResultShape, indicesElementType);
+
+  // Initialize the expanded output values
+  SmallVector<Value> dynSizes;
+  for (auto en : llvm::enumerate(valuesExpandedType.getShape())) {
+    if (en.value() == ShapedType::kDynamicSize) {
+      dynSizes.push_back(
+          rewriter.create<tensor::DimOp>(loc, valuesExpanded, en.index()));
+    }
+  }
+  Value initTensorOutputValues = rewriter.create<mlir::linalg::InitTensorOp>(
+      loc, dynSizes, outputValuesExpandedType.getShape(), valueElementType);
+  Value initTensorOutputIndices = rewriter.create<mlir::linalg::InitTensorOp>(
+      loc, dynSizes, outputIndicesExpandedType.getShape(), indicesElementType);
+
+  // Initialize indices to positive infinity and values to negative infinity
+  // for a top (maxk) comparison.
+  Attribute negInfAttr;
+  if (auto intType = valueElementType.dyn_cast<IntegerType>()) {
+    negInfAttr = rewriter.getIntegerAttr(
+        intType, APInt::getSignedMinValue(intType.getWidth()));
+  } else {
+    auto negApFloat =
+        APFloat::getInf(valueElementType.cast<FloatType>().getFloatSemantics(),
+                        /*Negative=*/true);
+    negInfAttr = rewriter.getFloatAttr(valueElementType, negApFloat);
+  }
+  Value negInf = rewriter.create<arith::ConstantOp>(loc, negInfAttr);
+  Attribute posInfAttr =
+      rewriter.getIntegerAttr(indicesElementType, APInt::getSignedMaxValue(32));
+  Value posInf = rewriter.create<arith::ConstantOp>(loc, posInfAttr);
+  Value negInfTensor =
+      rewriter.create<linalg::FillOp>(loc, negInf, initTensorOutputValues)
+          .result();
+  Value posInfTensor =
+      rewriter.create<linalg::FillOp>(loc, posInf, initTensorOutputIndices)
+          .result();
+
+  SmallVector<Type> parallelTopkResultTypes{outputValuesExpandedType,
+                                            outputIndicesExpandedType};
+  SmallVector<Value> parallelTopkIns{valuesExpanded};
+  SmallVector<Value> parallelTopkOuts{negInfTensor, posInfTensor};
```

Comment:
Prefer using "assignment-like syntax" or "traditional constructor syntax (with parentheses"

```suggestion
  SmallVector<Value> parallelTopkIns = {valuesExpanded};
  SmallVector<Value> parallelTopkOuts(negInfTensor, posInfTensor);
```


There are ambiguity in brace initialization. E.g.,

```
std::vector<std::string> strings{2}; // A vector of two empty strings.
std::vector<int> ints{2};            // A vector containing only the integer 2.